### ISSUE TITLE
Type-checking anomaly on constructor or method parameter passing

### ISSUE THREAD

**Issue by RobertoMalatesta** [Microsoft/TypeScript#10710]
**TypeScript Version:**  2.0.2

**Code**

// A *self-contained* demonstration of the problem follows...

```ts
//++++++++++++++++++++++++

enum FirstEnum { one=<any>"One", two=<any>"Two" }

type Extractor = (par:FirstEnum) => any[]

function IGetOnlyFirstEnums(par:FirstEnum){}

class X {
	constructor (public extractor:Extractor){}	
	public method01 (extractor:Extractor) {}
}

function extr (x:number):any[] {
	return []
}


IGetOnlyFirstEnums(extr) // OK: this gives type-check error as expected:
// error TS2345: Argument of type '(x: number) => any[]' is not assignable to parameter of type 'FirstEnum'.

let t = new X(extr)  // KO: this should give error but it does not.
t.method01(extr)   // KO: method as well is not issuing an error 

//++++++++++++++++++++++++

```

**Expected behavior:**

parameter-passing in both new and method call should raise an error each line: 
let t = new X(extr) 
t.method01(extr)   

**error TS2345: Argument of type '(x: number) => any[]' is not assignable to parameter of type** 
**error TS2345: Argument of type '(x: number) => any[]' is not assignable to parameter of type** 
just like calling the function does.

**Actual behavior:** 
No error is emitted for 
let t = new X(extr) 
t.method01(extr)   
`

--R

**Comment by mhegazy** [comment]
first the definition of `IGetOnlyFirstEnums` does not match the other two, I am guessing you meant `function IGetOnlyFirstEnums(par:Extractor){}`. if you do so, you should see the same behavior.

Not getting errors is what i would expect; Enums are subtype of number. and passing a function that can handle all numbers (`extr`) as a function that can handle a subset of it `FirstEnum`, is OK.

the problem really is you are hacking in a string enum using regular enums. the type system does not know you will be putting a string in the enum. you should be looking at https://github.com/Microsoft/TypeScript/issues/1206 and https://github.com/Microsoft/TypeScript/issues/1206

**Comment by RobertoMalatesta** [comment]
Hi @mhegazy ,
thanks for the answer.

I wrote this more  polished code that can be of help if someone has the same doubt:


```
enum FirstEnum { one=<any>"One", two=<any>"Two" }
enum SecondEnum {Lorem=<any>"Lorem", Ipsum=<any>"Ipsum"}


type CorrectExtractorType = (par:FirstEnum) => any[]
type wrongExtractorType = (par:SecondEnum) => any[]


function correctExtractor(par:FirstEnum)  {return []}
function wrongExtractor(par:SecondEnum)  {return []}
function genericNumExtractor(par:number)  {return []}

class X {
    constructor (public extractor:CorrectExtractorType){}  
    public method01 (extractor:CorrectExtractorType) {}
}



let c = new X(correctExtractor)   // OK
let e = new X(wrongExtractor)   // error
let g = new X(genericNumExtractor) // generic is OK

c.method01(wrongExtractor)  // error
c.method01(genericNumExtractor) // generic is OK
```


--R

**Comment by mhegazy** [comment]
To reiterate my earlier comment, do not use enums as a collection of strings.

**Comment by RobertoMalatesta** [comment]
Thank 4 the suggestion.This is more of a syntax inspection of what TS can do than actual use.
Anyway should be improved, even if ES compatibility is limiting.
